sched/adaptdl_sched/policy/pollux.py

```diff
@@ -311,17 +345,28 @@ def _mutation(self, states, **kwargs):
         prob = 1.0 / np.where(states > 0, num_nonzero, num_zero)
         prob = prob.reshape(states.shape)
         m = np.random.random(states.shape) < prob
-        r = np.random.randint(np.iinfo(np.int16).max, size=states.shape)
+        r = np.random.randint(self._min_replicas, self._max_replicas + 1,
+                              size=states.shape)
```

Comment:
Is this change necessary? Since it impacts the current behavior for preemptible jobs too.

Reply:
So according to [this](https://view.officeapps.live.com/op/view.aspx?src=http%3a%2f%2fvideo.ch9.ms%2fsessions%2fgonat%2f2013%2fSTLGN13rand.pptx) `rand() % 100` does not guarantee uniform random integers between range [0, 100). Also we have the exact range available which should guarantee a uniform distribution.

Reply:
But I don't think what you changed it to is uniformly at random either. For example in the edge case that `min_replicas == max_replicas` then `r` may only take on a single value. The original code is approximately uniform in the case that `max_replicas ` << 2^16, which should certainly be true.

Reply:
But why would you even want randomness for the edge case of `min_replicas = max_replicas`? There are no (better) solutions to explore, the condition is the optimal solution. We just have to execute it.

Reply:
Ok, bad example. Suppose `min_replicas=4`, `max_replicas=8` and currently `state=6`. `r` will be uniformly one of 4, 5, 6, 7, or 8, which means after the next line, `state` will have a new value of 1, 2, 3, 4, or 5.

I also just realized lines 355-358 will just hard-set `state=4` anyways, which certainly is not uniformly at random between `min_replicas=4` and `max_replicas=8`.

Reply:
yes that hard setting is addressing the specific scalability issue that we found with very large values of `min_replicas`. You can think of it as a bias we add to the random value. uniform `r` applies to all other jobs with `min_replicas=0`.